### ISSUE TITLE
ldap: finished compiling filter with extra at end: 

### ISSUE THREAD

**Issue by hridyeshpant** [go-ldap/ldap#56]
Hi,
I am getting below error which authenticating ldap ,could some one please help me why i am this error?

ldap: finished compiling filter with extra at end: `OU=User Policy 0,OU=All Users,DC=**,DC=**,DC=**,DC=com)(uniqueMember=CN=user,OU=User Policy 0,OU=All Users,DC=**,DC=**,DC=**,DC=com))`

but when i trying to use ldap for user which is in different tree like below ,it working fine
dn: `CN=user2,OU=Service Accounts,OU=Non-User,DC=**,DC=**,DC=**,DC=com`

Does string `OU=User Policy 0` is causing issue in first case?
https://github.com/go-ldap/ldap/search?utf8=%E2%9C%93&q=extra

**Comment by liggitt** [comment]
do you have the code snippet you are using? the example filter looks like it has mismatched parenthesis

**Comment by dgorst** [comment]
Hey I'm seeing the same results here - some users work ok, others do not.

For reference, I'm using Hashicorp Vault which under the hood is using this LDAP library.

If it provides a clue this is an obfuscated lookup of my own username:

```
$ ldapsearch sAMAccountName=jbloggs
SASL/GSSAPI authentication started
SASL username: jbloggs@ABC.CORP.ABCD.COM
# extended LDIF
#
# LDAPv3
# base <> (default) with scope subtree
# filter: sAMAccountName=jbloggs
# requesting: ALL
#

# Jo Bloggs (jbloggs), User Policy 0, All Users, ABC.CORP.ABCD.com
dn: CN=Jo Bloggs (jbloggs),OU=User Policy 0,OU=All Users,DC=ABC,DC=CORP,DC=ABCD,DC=com
objectClass: top
objectClass: person
objectClass: organizationalPerson
objectClass: user
cn: Jo Bloggs (jbloggs)
sn: Bloggs
```

The error seems to occur in `/ldap/backend.go` in Vault:

```
	// Enumerate all groups the user is member of. The search filter should
	// work with both openldap and MS AD standard schemas.
	sresult, err := c.Search(&ldap.SearchRequest{
		BaseDN: cfg.GroupDN,
		Scope:  2, // subtree
		Filter: fmt.Sprintf("(|(memberUid=%s)(member=%s)(uniqueMember=%s))", ldap.EscapeFilter(username), ldap.EscapeFilter(userdn), ldap.EscapeFilter(userdn)),
	})
	if err != nil {
		return nil, logical.ErrorResponse(fmt.Sprintf("LDAP search failed: %v", err)), nil
	}
```

The resulting error is:

```
* LDAP search failed: LDAP Result Code 201 "": ldap: finished compiling filter with extra at end: OU=User Policy 0,OU=All Users,DC=ABC,DC=CORP,DC=ABCD,DC=com)(uniqueMember=CN=Jo Bloggs (jbloggs),OU=User Policy 0,OU=All Users,DC=ABC,DC=CORP,DC=ABCD,DC=com))
```

In addition though, here's a user which DOES work OK:

```
dn: CN=s-robot,OU=Service Accounts,OU=Non-User,DC=ABC,DC=CORP,DC=ABCD,DC=com
```

Any ideas? At first glance it seems like the parenthesis in my `dn` value is confusing something but I'm probably wrong there - it's the only real difference between the two accounts!

**Comment by dgorst** [comment]
Ok, had a closer look and added a couple of tests to `filter_test.go` that reproduce the queries that Vault is producing - the first is the above DN without parenthesis, the second with:

```
	compileTest{
		filterStr:      "(|(memberUid=JBLOGGS)(member=CN=Jo Bloggs jbloggs,OU=User Policy 0,OU=All Users,DC=ABC,DC=CORP,DC=ABCD,DC=com)(uniqueMember=CN=Jo Bloggs jbloggs,OU=User Policy 0,OU=All Users,DC=ABC,DC=CORP,DC=ABCD,DC=com))",
		expectedFilter: "(|(memberUid=JBLOGGS)(member=CN=Jo Bloggs jbloggs,OU=User Policy 0,OU=All Users,DC=ABC,DC=CORP,DC=ABCD,DC=com)(uniqueMember=CN=Jo Bloggs jbloggs,OU=User Policy 0,OU=All Users,DC=ABC,DC=CORP,DC=ABCD,DC=com))",
		expectedType: 	ldap.FilterOr,
	},
	compileTest{
		filterStr:      "(|(memberUid=JBLOGGS)(member=CN=Jo Bloggs (jbloggs),OU=User Policy 0,OU=All Users,DC=ABC,DC=CORP,DC=ABCD,DC=com)(uniqueMember=CN=Jo Bloggs (jbloggs),OU=User Policy 0,OU=All Users,DC=ABC,DC=CORP,DC=ABCD,DC=com))",
		expectedFilter: "(|(memberUid=JBLOGGS)(member=CN=Jo Bloggs (jbloggs),OU=User Policy 0,OU=All Users,DC=ABC,DC=CORP,DC=ABCD,DC=com)(uniqueMember=CN=Jo Bloggs (jbloggs),OU=User Policy 0,OU=All Users,DC=ABC,DC=CORP,DC=ABCD,DC=com))",
		expectedType: 	0,
		expectedErr:    "finished compiling filter with extra at end",
	},
```

This seems to confirm the issue is with the parenthesis in the DN.

Taking a closer look...

**Comment by dgorst** [comment]
You know what, I think this ldap library is fine and the problem lied within Vault itself - this commit from just over a month ago fixes the issue I'm seeing and is available in Vault 0.5.2:

https://github.com/hashicorp/vault/commit/a2aad0bbd6fa5ff3148d9ab9a5ee3838d0b0985d

Completely replicating the latest Vault code in `filter_test.go` works fine by the way:

```
	compileTest{
		filterStr:      fmt.Sprintf("(|(memberUid=%s)(member=%s)(uniqueMember=%s))", ldap.EscapeFilter("JBLOGGS"), ldap.EscapeFilter("CN=Jo Bloggs (jbloggs),OU=User Policy 0,OU=All Users,DC=ABC,DC=CORP,DC=ABCD,DC=com"), ldap.EscapeFilter("CN=Jo Bloggs (jbloggs),OU=User Policy 0,OU=All Users,DC=ABC,DC=CORP,DC=ABCD,DC=com")),
		expectedFilter: "(|(memberUid=JBLOGGS)(member=CN=Jo Bloggs \\28jbloggs\\29,OU=User Policy 0,OU=All Users,DC=ABC,DC=CORP,DC=ABCD,DC=com)(uniqueMember=CN=Jo Bloggs \\28jbloggs\\29,OU=User Policy 0,OU=All Users,DC=ABC,DC=CORP,DC=ABCD,DC=com))",
		expectedType: 	ldap.FilterOr,
	},
```

Not seeing the above issue anymore...
